metriq_gym/benchmarks/clops.py

```diff
@@ -0,0 +1,155 @@
+from metriq_gym.helpers.execution_time import execution_time
+from metriq_gym.helpers.task_helpers import flatten_job_ids
+from metriq_gym.helpers.topology_helpers import device_topology
+
+
+@dataclass
+class ClopsData(BenchmarkData):
+    pass
+
+
+@dataclass
+class ClopsResult(BenchmarkResult):
+    clops_score: float
+
+
+def create_qubit_map(width, topology_graph: rx.PyGraph, faulty_qubits, total_qubits):
+    """
+    Returns a list of  'width' qubits that are connected based on a coupling map that has
+    bad edges already removed and a list of faulty qubits.  If there is not
+    such a map, raised ValueError
+    """
+    qubit_map = []
+
+    # make coupling map bidirectional so we can find all neighbors
+    cm = copy.deepcopy(topology_graph)
+    for edge in topology_graph.edge_list():
+        cm.add_edge(edge[1], edge[0], None)
+
+    for starting_qubit in range(total_qubits):
+        if starting_qubit not in faulty_qubits:
+            qubit_map = [starting_qubit]
+            new_neighbors = []
+            prospective_neighbors = list(cm.neighbors(starting_qubit))
+            while prospective_neighbors:
+                for pn in prospective_neighbors:
+                    if pn not in qubit_map and pn not in faulty_qubits:
+                        new_neighbors.append(pn)
+                qubit_map = qubit_map + new_neighbors
+                prospective_neighbors = []
+                for nn in new_neighbors:
+                    potential_neighbors = list(cm.neighbors(nn))
+                    for pn in potential_neighbors:
+                        if pn not in prospective_neighbors:
+                            prospective_neighbors.append(pn)
+                new_neighbors = []
+                if len(qubit_map) >= width:
+                    return qubit_map[:width]
+    raise ValueError("Insufficient connected qubits to create set of %d qubits", width)
+
+
+# adapted from submodules/qiskit-device-benchmarking/qiskit_device_benchmarking/clops/clops_benchmark.py::prepare_clops_circuits
+# As opposed to the original version of this function, this takes a topology graph as an argument instead of an IBM's coupling map object.
+def append_2q_layer(
+    qc: QuantumCircuit, topology_graph: rx.PyGraph, basis_gates: list[str], rng
+) -> None:
+    """
+    Add a layer of random 2q gates.
+    """
+    available_edges = set(topology_graph.edge_list())
+    while len(available_edges) > 0:
+        edge = tuple(rng.choice(list(available_edges)))
+        available_edges.remove(edge)
+        edges_to_delete = []
+        for ce in list(available_edges):
+            if (edge[0] in ce) or (edge[1] in ce):
+                edges_to_delete.append(ce)
+        available_edges.difference_update(set(edges_to_delete))
+        if "ecr" in basis_gates:
+            qc.ecr(*edge)
+        elif "cz" in basis_gates:
+            qc.cz(*edge)
+        else:
+            qc.cx(*edge)
+    return
```

Comment:
Technically, not putting `return` has the same behavior as returning `None`, so you can remove this `return` statement as it is redundant (all functions in Python return None by default unless specified otherwise).